### ISSUE TITLE
Distribute react-relay and other packages as ESM

### ISSUE THREAD

**Issue by gajus** [facebook/relay#4548]
Feels like this change is well overdue.

**Comment by captbaritone** [comment]
I agree. This would be a welcome improvement. When setting up Relay for side projects I often hit rough edges because of this.

If you, or anyone else, wants to start collecting a full picture of what work would be required here, that would be very helpful. Some questions on my mind:

* Is it a prerequisite to move to ESM within Relay? Currently we use commonJS require/module.exports everywhere. Internally, we've long wanted to move to ES6 imports/exports but simply haven't had the bandwidth to prioritize it.
* What changes would we need to make to our build pipeline? Is it possible with the tools we have today?
* Can we support both commonJS and ESM? Will this be a breaking/problematic change for some users?

**Comment by gajus** [comment]
Highly theoretically, you could just release `relay-esm` package that simply re-exports everything as ESM:

```ts
// Because Relay is a CommonJS module, we are having problems with Vite bundling it.
// Until Relay becomes an ESM module, we need to use this package to re-export it.
import { default as reactRelay } from 'react-relay';
import { default as relayRuntime } from 'relay-runtime';

export { type PreloadedQuery, type UseMutationConfig } from 'react-relay';
export {
  type CacheConfig,
  type Disposable,
  type FetchPolicy,
  type FragmentRefs,
  type GraphQLResponse,
  type GraphQLTaggedNode,
  type MutationParameters,
  type OperationType,
  type PayloadError,
  type RequestParameters,
  type SelectorStoreUpdater,
  type Variables,
  type VariablesOf,
} from 'relay-runtime';

export const ConnectionHandler = relayRuntime.ConnectionHandler;
export const Environment = relayRuntime.Environment;
export const Network = relayRuntime.Network;
export const QueryResponseCache = relayRuntime.QueryResponseCache;
export const RecordSource = relayRuntime.RecordSource;
export const Store = relayRuntime.Store;

export const loadQuery = reactRelay.loadQuery;
export const fetchQuery = reactRelay.fetchQuery;
export const useLazyLoadQuery = reactRelay.useLazyLoadQuery;
export const usePreloadedQuery = reactRelay.usePreloadedQuery;
export const graphql = reactRelay.graphql;
export const useFragment = reactRelay.useFragment;
export const useMutation = reactRelay.useMutation;
export const usePaginationFragment = reactRelay.usePaginationFragment;
export const useRelayEnvironment = reactRelay.useRelayEnvironment;
export const readInlineData = reactRelay.readInlineData;
export const useRefetchableFragment = reactRelay.useRefetchableFragment;
export const RelayEnvironmentProvider = reactRelay.RelayEnvironmentProvider;
export const commitMutation = reactRelay.commitMutation;
```

That's what we do internally in Contra.

**Comment by matclayton** [comment]
We also keep bumping into this edge case, now with ESM support for GraphQL codegen, it feels like this would be very welcome and more consistent/beneficial for new comers to Relay

**Comment by orta** [comment]
This had bite me when dealing with Vite v5+ and server side rendering. For folks exploring the answer, I found using [vite-plugin-cjs-interop](https://www.npmjs.com/package/vite-plugin-cjs-interop) was enough to get everything working - https://github.com/orta/relay-vite-ssr-example/commit/181baccc47589fe6756d8956e63e72fe73053272

I'm afraid that I don't have any use on the actual structural answers to @captbaritone's questions, a lot of the hard integration work for ESM support in TS came after I left. ESM is still a bit of a mystery to me.

**Comment by matclayton** [comment]
@orta we had the exact same solution, it’s mainly a wart in the ssr build chain and feels like it’ll break at some point in an upgrade to Vite.

**Comment by jantimon** [comment]
To properly ship ESM we should simplify the babel compilation process and so we keep only flow type stripping (`@babel/plugin-transform-flow-strip-types`)

Here's what the source actually uses today (excluding tests, ignoring Flow types and import/export which always need transformation):

| Feature | Count |
|---|---|
| Arrow functions | ~960 |
| Template literals | ~943 |
| Object spread | ~819 |
| Destructuring | ~405 |
| Optional chaining (`?.`) | ~240 |
| Nullish coalescing (`??`) | ~184 |
| `for...of` | ~137 |
| `class` / `extends` | ~36 |
| `async`/`await` | 2-3 experimental files |
| `Map` / `Set` | extensively |

`?.` is the binding constraint. The minimum targets to ship all of the above untranspiled:

| Browser | Min Version | Released |
|---|---|---|
| Chrome | 80+ | Feb 2020 |
| Firefox | 74+ | Mar 2020 |
| Safari | 13.1+ | Mar 2020 |
| Node.js | 14+ | Apr 2020 |

Without the transpilation the source doesn't use any built-ins that need runtime polyfills beyond what these browsers support natively  **so `@babel/plugin-transform-runtime` (CJS!) and `@babel/runtime` (also CJS!) could be dropped entirely as well**

For apps that still need to support older browsers can handle transpilation themselves via their own bundler config.

For me the idea would be to:

1. change `module.exports` to `export` inside all browser targeting source files
2. change `require` to `import`  inside all browser targeting source files
3. remove all outdated babel presets and transforms
4. add dual ESM+CJS compilation output
5. change package.json for react-relay and relay-runtime to `"type": "module"` with correct export fields (we can keep CJS for backwards compatibility)
6. pick up my old PR (sorry for not doing it earlier) #4873 so that the relay rust compiler is compatible too
What do you think?

**Comment by jantimon** [comment]
@captbaritone the `injectPromisePolyfill` is setTimeout based which makes it really hard to migrate the tests to a leaner compilation flow

If we clean that up we can get rid of inline-requires in the test preprocessor (only reason we need it is to ensure resolveImmediate captures the polyfilled Promise)

https://github.com/facebook/relay/blob/36fb56094ada8c7e206e268d7aef8c33872b5e5b/packages/relay-test-utils-internal/injectPromisePolyfill__DEPRECATED.js#L13-L33

can we remove the `injectPromisePolyfill` first to make the migration to ESM easier?

**Comment by jantimon** [comment]
I prepared #5209 (which adds ESM for the rust based relay compiler) and #5210 (which removes the deprecated `injectPromisePolyfill`)

Please let me know what you think

The next step would be to simplify the build chain and add dual ESM+CJS compilation
